### PR TITLE
Refactor hook plugin config loading, expose configuration for label plugin

### PR DESCRIPTION
**Pull request by stevekuznetsov**
vendor: ghodss/yaml: update to 0ca9ea5df5451ffdf184b4428c902747c2c11cd7

Update to latest ghodss/yaml version to allow for unmarshalling into
structs containing struct pointers.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Bring hook plugin configuration into plugins.yaml

As more and more plugins gain configuration options, it is becoming
clear that the overall Prow configuration is not the appropriate place
for these options. Configuration options for these plugins are almost
always created in order to support more organizations or projects, so
the mapping of project --> plugins we already have in config.yaml is
the correct place to put these options. With this patch, that config
file now maps project --> plugins and config. Config for the repository
being operated on is exposed to handlers at run-time.

---

Configure organization for SIG labels

Supporting SIG labels for SIGs in other organizations requires that the
label-matching regular expression be configurable with the name of the
organization that owns SIGs.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

/area prow
/cc @spxtr @kargakis I was trying to configure the `label` plugin... but as per the older discussion, adding everything into the global Prow config was getting ugly so I took a stab at refactoring that mechanism.

### CODE REVIEW

File: prow/plugins/plugins.go [hunk 1]
```diff
@@ -107,54 +110,154 @@ type PluginAgent struct {
...
+
 // Load attempts to load config from the path. It returns an error if either
 // the file can't be read or it contains an unknown plugin.
 func (pa *PluginAgent) Load(path string) error {
-	b, err := ioutil.ReadFile(path)
+	reader, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	// We want to deserialize to two different targets
```
Comment (stevekuznetsov) [inline prow/plugins/plugins.go]: This really sucks but I wanted to keep the same API for plugin lists as we had before... and Go has no simple way to do strict JSON deserialization.

File: prow/plugins/plugins.go [hunk 2]
```diff
@@ -279,13 +382,58 @@ func (pa *PluginAgent) PushEventHandlers(owner, repo string) map[string]PushEven
...
+	} else if orgOk && repoOk {
+		return mergeConfig(orgConfig, repoConfig)
+	}
+
+	panic(fmt.Sprintf("No configuration was found for %s!", fullName))
+}
+
+// mergeConfig merges organization and repository plugin config
+// to create the full configuration for a set of repository plugins.
+func mergeConfig(org, repo Config) Config {
```
Comment (stevekuznetsov) [inline prow/plugins/plugins.go]: Instead, we could resolve configuration at the start if we know the set of repos we will be looking at

File: prow/plugins/plugins.go [hunk 3]
```diff
@@ -44,49 +47,49 @@ var (
 	statusEventHandlers        = map[string]StatusEventHandler{}
 )
 
-type IssueHandler func(PluginClient, github.IssueEvent) error
+type IssueHandler func(PluginClient, Config, github.IssueEvent) error
```
Comment (spxtr) [inline prow/plugins/plugins.go]: I feel like this Config should be a member of the PluginClient.

Comment (stevekuznetsov) [inline prow/plugins/plugins.go]: I wasn't sure if it quite fit there, as the plugins would then need to resolve which config by looking at their own repo and org? That seemed like something that should happen before the registered handlers were called

**Comment by spxtr** [comment]
I'm not a huge fan of the change to the plugin config YAML. We go from
```yaml
triggers:		
- repos:		
  - kubernetes		
  - kubernetes-incubator		
  - kubernetes-security		
  - google/cadvisor		
  trusted_org: kubernetes
```
to
```yaml
google/cadvisor:
  trigger:
    trusted_org: kubernetes

kubernetes/charts:
  trigger:
    trusted_org: kubernetes

kubernetes/heapster:
  trigger:
    trusted_org: kubernetes

kubernetes/kops:
  trigger:
    trusted_org: kubernetes

kubernetes/kubernetes:
  trigger:
    trusted_org: kubernetes

kubernetes/test-infra:
  trigger:
    trusted_org: kubernetes
```

I like that the original neatly describes that these repos are related. Could we keep the separation between "these plugins are enabled on these repos" and "here's how to configure this plugin on these repos," but just move the latter into `plugins.yaml`?

**Comment by kargakis** [comment]
> I like that the original neatly describes that these repos are related. Could we keep the separation between "these plugins are enabled on these repos" and "here's how to configure this plugin on these repos," but just move the latter into plugins.yaml?

I think I agree with this. While we lose fine-grained plugin tuning, in most cases I can think of, the configuration for the existing plugins will be the same across all repos.

**Comment by stevekuznetsov** [comment]
Yeah, that part is definitely not the best. If, however, we can move the `trigger` configuration into the `kubernetes` org-level config, we are better off on both fronts. The `trigger` config was actually why I decided to encode plugin config as a mapping of repo-->config ... otherwise, the `plugins.yaml` would have more than one way of mapping repo name --> plugins enabled and repo name --> plugins configured which seems bad as well.

**Comment by stevekuznetsov** [comment]
> If, however, we can move the `trigger` configuration into the `kubernetes` org-level config

Which I did not do as I was not really sure the implication here -- are we munging repos in the `kubernetes` org that there are no `triggers` for?

**Comment by stevekuznetsov** [comment]
To confirm -- today for _e.g._ `kubernetes/heapster` do we just trust everyone for the triggers there?

**Comment by spxtr** [comment]
> To confirm -- today for _e.g._ `kubernetes/heapster` do we just trust everyone for the triggers there?

No, because it's under the `kubernetes` org we use the `trusted_org: kubernetes` config line.

**Comment by stevekuznetsov** [comment]
@spxtr so then I've done the conversion wrong and we should have gone from:

```yaml
triggers:
- repos:
  - kubernetes
  - kubernetes-incubator
  - kubernetes-security
  - google/cadvisor
  trusted_org: kubernetes
```

To:
```yaml
kubernetes:
  trigger:
    trusted_org: kubernetes

kubernetes-incubator:
  trigger:
    trusted_org: kubernetes

kubernetes-security:
  trigger:
    trusted_org: kubernetes

google/cadvisor:
  trigger:
    trusted_org: kubernetes
```

@kargakis @spxtr re: not needing repo-specific config -- do we expect different top-level organizations to share SIGs between each other?

File: prow/hook/events.go [hunk 4]
```diff
@@ -33,9 +33,10 @@ func (s *Server) handleReviewEvent(re github.ReviewEvent) {
 		"url":      re.Review.HTMLURL,
 	})
 	l.Infof("Review %s.", re.Action)
-	for p, h := range s.Plugins.ReviewEventHandlers(re.PullRequest.Base.Repo.Owner.Login, re.PullRequest.Base.Repo.Name) {
+	org, repo := re.PullRequest.Base.Repo.Owner.Login, re.PullRequest.Base.Repo.Name
+	for p, h := range s.Plugins.ReviewEventHandlers(org, repo) {
 		go func(p string, h plugins.ReviewEventHandler) {
-			pc := s.Plugins.PluginClient
+			pc := s.pluginClientFor(org, repo)
```
Comment (stevekuznetsov) [inline prow/hook/events.go]: This is not thread safe ... but the issue here is that right now we do org/repo resolution here but then the `trigger` plugin needs to do the resolution again -- lots of room for error there if it resolves org/repo incorrectly or differently from here. Hard to have a global read-only plugin config that would be different for each of these calls and resolved here. I can keep the status quo but it doesn't seem right

Comment (stevekuznetsov) [inline prow/hook/events.go]: @spxtr let me know what you think -- if you think we should just keep the resolution strategy the same I will not try to work around it

**Comment by kargakis** [comment]
> @kargakis @spxtr re: not needing repo-specific config -- do we expect different top-level organizations to share SIGs between each other?

Possible. See the existing organizations. SIGs work on different repositories across all orgs.

**Comment by spxtr** [comment]
https://github.com/kubernetes/test-infra/pull/3790#issuecomment-319433845

Actually these are different. In the latter case the trigger plugin will be running against all kubernetes repos, whereas in the former it may not be. This is actually important, because it will make the "ok to test" comment even if there are no tests to trigger. We can fix that, but it's still a difference.

**Comment by stevekuznetsov** [comment]
> We can fix that, but it's still a difference.

@spxtr I see. I've reverted the config changes so the plugins have the same global config entries they had before, just exposed through the other file.

File: prow/plugins.yaml [hunk 5]
```diff
@@ -2,50 +2,63 @@
 # Keys: Full repo name: "org/repo".
 # Values: List of plugins to run against the repo.
 ---
-google/cadvisor:
```
Comment (kargakis) [inline prow/plugins.yaml]: Do we need to touch this? It's harder to review what's added.

Comment (stevekuznetsov) [inline prow/plugins.yaml]: Not sure how to get a sane deserialization with config and plugins in the same object -- no good way to do a `map[string][]string` that "soaks up" the rest of the keys that did not get used for first-class deserialization AFAICT

Comment (kargakis) [inline prow/plugins.yaml]: Can you get a commit that moves from the old to the new schema and a follow-up with the new config?

Comment (stevekuznetsov) [inline prow/plugins.yaml]: Not sure what you mean -- the two changes are:

1. take entirety of current `plugins.yaml`, indent it once and call it `plugins:`
2. copy-paste plugin config from `config.yaml` into `plugins.yaml` at the top level

**Comment by kargakis** [comment]
Don't you need to update Godeps.json?

**Comment by stevekuznetsov** [comment]
@kargakis whoops, not sure how that slipped past

**Comment by stevekuznetsov** [comment]
[(1) [k8sCI] Enable the label plugin in prow](https://trello.com/c/FM7RZkhi/267-1-k8sci-enable-the-label-plugin-in-prow)

File: prow/hook/events.go (hunk 4 above)
Comment (stevekuznetsov) [inline prow/hook/events.go]: @spxtr @kargakis thoughts on this?

File: prow/hook/events.go [hunk 6]
```diff
@@ -171,3 +178,12 @@ func (s *Server) handleStatusEvent(se github.StatusEvent) {
 		}(p, h)
 	}
 }
+
+func (s *Server) pluginClientFor(org, repo string) plugins.PluginClient {
+	client := s.Plugins.PluginClient
+	if trigger := s.Plugins.TriggerFor(org, repo); trigger != nil && trigger.TrustedOrg != "" {
```
Comment (stevekuznetsov) [inline prow/hook/events.go]: See my above comment on that... not happy about it either: https://github.com/kubernetes/test-infra/pull/3790#discussion_r130691307

File: prow/hook/events.go (hunk 4 above)
Comment (stevekuznetsov) [inline prow/hook/events.go]: @spxtr @kargakis thoughts here? should I just revert to the current behavior>

Comment (spxtr) [reply prow/hook/events.go]: I think it makes sense to revert to current behavior and then try to make sure that the various plugins that need to think about "org/repo" strings use the same code to do so. We can add a simple package that just checks if a given repo is under a given org with simple string matching.

**Comment by stevekuznetsov** [comment]
Migrated to #4213

**Pull request closed without merging**